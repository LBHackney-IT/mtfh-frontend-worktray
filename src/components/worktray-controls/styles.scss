.mtfh-worktray-controls {
  &__controls {
    display: flex;
    justify-content: flex-end;
    flex-wrap: wrap;
    margin-top: 0;
  }

  .lbh-form-group {
    display: flex;
    align-items: center;
    margin-top: 0;

    .lbh-label {
      margin-right: 1rem;
      margin-bottom: 0;
    }

    & + .lbh-form-group {
      margin-left: 1rem;
    }
  }
}
