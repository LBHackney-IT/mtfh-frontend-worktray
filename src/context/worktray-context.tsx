import React, {
  Dispatch,
  FC,
  Reducer,
  createContext,
  useContext,
  useEffect,
  useMemo,
  useReducer,
} from "react";
import { useHistory, useLocation } from "react-router-dom";

import { stringify } from "query-string";

import { AxiosSWRError, useAxiosSWR } from "@mtfh/common/lib/http";

import { config } from "../services";
import {
  LimitOptions,
  OrderByOptions,
  Process,
  ProcessSortOptions,
  TimePeriodOptions,
  WorktrayFilterOptions,
} from "../types";

interface WorktrayResults {
  results: {
    processes: Process[];
  };
  total: number;
}

export type WorktrayState = {
  page: number;
  pageSize: LimitOptions;
  timePeriod: TimePeriodOptions;
  sort: ProcessSortOptions;
  order: OrderByOptions;
  patch?: string;
  areaId?: string | null;
  processNames?: string;
  status?: string;
  results?: Process[];
  total?: number;
  error?: AxiosSWRError;
};

export type WorktrayActions =
  | {
      type: "PAGE";
      payload: number;
    }
  | {
      type: "LIMIT";
      payload: LimitOptions;
    }
  | {
      type: "TIME_PERIOD";
      payload: TimePeriodOptions;
    }
  | {
      type: "SORT";
      payload: ProcessSortOptions;
    }
  | {
      type: "FILTER";
      payload: {
        type: WorktrayFilterOptions;
        payload: string;
      };
    }
  | {
      type: "ORDER";
      payload: OrderByOptions;
    };

export interface WorktrayProviderProps {
  initial: Partial<WorktrayState>;
}

const getInitialState = (
  options: WorktrayProviderProps["initial"] = {},
): WorktrayState => {
  return {
    page: options.page || 1,
    pageSize: options.pageSize || LimitOptions.SMALL,
    timePeriod: options.timePeriod || TimePeriodOptions.DAYS_30,
    sort: options.sort || ProcessSortOptions.STATUS,
    order: options.order || OrderByOptions.ASC,
    processNames: options.processNames || undefined,
    patch: options.patch || undefined,
    areaId: options.areaId || undefined,
    status: options.status || undefined,
  };
};

export const WorktrayContext = createContext<{
  state: WorktrayState;
  dispatch: Dispatch<WorktrayActions>;
}>({
  state: getInitialState(),
  dispatch: () => {},
});

export const WorktrayProvider: FC<WorktrayProviderProps> = ({ children, initial }) => {
  const reducer: Reducer<WorktrayState, WorktrayActions> = (state, action) => {
    switch (action.type) {
      case "PAGE": {
        return {
          ...state,
          page: action.payload,
        };
      }
      case "LIMIT": {
        if (Object.values(LimitOptions).includes(action.payload)) {
          return {
            ...state,
            pageSize: action.payload,
          };
        }
        return state;
      }
      case "TIME_PERIOD": {
        if (Object.values(TimePeriodOptions).includes(Number(action.payload))) {
          return {
            ...state,
            timePeriod: action.payload,
          };
        }
        return state;
      }
      case "SORT": {
        if (Object.values(ProcessSortOptions).includes(action.payload)) {
          return {
            ...state,
            sort: action.payload,
          };
        }
        return state;
      }
      case "ORDER": {
        if (Object.values(OrderByOptions).includes(action.payload)) {
          return {
            ...state,
            order: action.payload,
          };
        }
        return state;
      }
      case "FILTER": {
        return {
          ...state,
          [action.payload.type]: action.payload.payload,
        };
      }
      default: {
        return state;
      }
    }
  };

  const initialState = getInitialState(initial);
  const [state, dispatch] = useReducer(reducer, initialState);
  const { ...query } = state;

  const { data, error } = useAxiosSWR<WorktrayResults>(
    `${config.searchApiUrl}/search/processes?${stringify(
      {
        searchText: query.patch,
        page: query.page,
        pageSize: query.pageSize,
        timePeriod: query.timePeriod,
        processNames: query.processNames,
        status: query.status,
        sortBy: query.sort,
        isDesc: query.order === "asc",
      },
      {
        skipEmptyString: true,
      },
    )}`,
    {
      refreshWhenHidden: false,
      revalidateOnFocus: false,
    },
  );

  useEffect(() => {
    if (data?.total && data.total > 0) {
      const totalPages = Math.ceil(data.total / state.pageSize);
      if (state.page > totalPages) {
        dispatch({ type: "PAGE", payload: totalPages });
      }
    }
  }, [state.page, data?.total, state.pageSize]);

  const value = useMemo(
    () => ({
      state: {
        ...state,
        results: data?.results.processes,
        total: data?.total,
        error,
      },
      dispatch,
    }),
    [state, data, error, dispatch],
  );

  return <WorktrayContext.Provider value={value}>{children}</WorktrayContext.Provider>;
};

export const PersistWorktrayContextURL = ({
  sessionKey,
}: {
  sessionKey?: string;
}): null => {
  const {
    state: { page, pageSize, order, sort, timePeriod, patch, processNames, status },
  } = useContext(WorktrayContext);
  const { push } = useHistory();

  useEffect(() => {
    const query = new URLSearchParams();
    if (page > 1) {
      query.set("p", String(page));
    }

    if (pageSize !== LimitOptions.SMALL) {
      query.set("l", String(pageSize));
    }

    query.set("t", String(timePeriod));

    if (order && order !== OrderByOptions.ASC) {
      query.set("o", order);
    }

    query.set("sort", String(sort));

    if (patch) {
      query.set("patch", patch);
    }

    if (processNames) {
      query.set("processNames", processNames);
    }

    if (status) {
      query.set("status", status);
    }

    const path = `?${query.toString()}`;
    push(path);

    if (sessionKey) {
      window.sessionStorage.setItem(sessionKey, path);
    }
  }, [
    page,
    pageSize,
    order,
    sort,
    sessionKey,
    push,
    timePeriod,
    patch,
    processNames,
    status,
  ]);

  return null;
};

export const WorktrayURLProvider: FC<{
  sessionKey?: string;
  patchId?: string | null;
  areaId?: string | null;
}> = ({ children, sessionKey, patchId, areaId }): JSX.Element => {
  const { search } = useLocation();
  const query = new URLSearchParams(search);
  const page = Number(query.get("p") || 1);
  const pageSize = Number(query.get("l") || 10);
  const timePeriod = (query.get("t") || "") as TimePeriodOptions;
  const patch = query.get("patch") || patchId || "";
  const processNames = query.get("processNames") || "";
  const status = query.get("status") || "";
  const sortBy = query.get("sort");
  const orderBy = query.get("o");

  const sort = useMemo(() => {
    if (sortBy === "name") {
      return ProcessSortOptions.NAME;
    }
    return ProcessSortOptions.STATUS;
  }, [sortBy]);

  const order = useMemo(() => {
    if (orderBy === "desc") {
      return OrderByOptions.DESC;
    }
    return OrderByOptions.ASC;
  }, [orderBy]);

  return (
    <WorktrayProvider
      initial={{
        page,
        sort,
        order,
        pageSize,
        timePeriod,
        patch,
        areaId,
        processNames,
        status,
      }}
    >
      {children}
      <PersistWorktrayContextURL sessionKey={sessionKey} />
    </WorktrayProvider>
  );
};
