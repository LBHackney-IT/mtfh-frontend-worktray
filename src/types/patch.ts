export interface ResponsibleEntity {
  id: string;
  name: string;
  responsibleType: string;
}

export interface Patch {
  id: string;
  parentId: string;
  name: string;
  patchType: string;
  domain: string;
  responsibleEntities: ResponsibleEntity[];
}
