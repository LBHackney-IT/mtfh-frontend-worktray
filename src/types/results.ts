import { ProcessState } from "@mtfh/common/lib/api/process/v1";

export interface WorktrayResult {
  id: string;
  processName: string;
  targetType: string;
  targetId: string;
  currentState: ProcessState;
  relatedEntities: Record<string, string>[];
}

export enum ProcessSortOrderOptions {
  STATUS = "status",
  TIME_LEFT_ASC = "time-left-asc",
  TIME_LEFT_DESC = "time-left-desc",
  STATE_ASC = "state-asc",
  STATE_DESC = "state-desc",
  PATCH_ASC = "patch-asc",
  PATCH_DESC = "patch-desc",
  PROCESS_ASC = "process-asc",
  PROCESS_DESC = "process-desc",
  NAME_ASC = "name-asc",
  NAME_DESC = "name-desc",
  ADDRESS_ASC = "address-asc",
  ADDRESS_DESC = "address-desc",
}

export enum ProcessSortOptions {
  STATUS = "status",
  TIME_LEFT = "time_left",
  STATE = "state",
  PATCH = "patch",
  PROCESS = "process",
  NAME = "name",
  ADDRESS = "address",
}

export enum OrderByOptions {
  ASC = "asc",
  DESC = "desc",
}

export enum LimitOptions {
  SMALL = 12,
  MEDIUM = 20,
  LARGE = 40,
}
