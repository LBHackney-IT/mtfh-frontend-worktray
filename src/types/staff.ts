export interface StaffResult {
  firstName: string;
  lastName: string;
  emailAddress: string;
  patchId: string | null;
}

export interface StaffResults {
  results: {
    staff: StaffResult[];
  };
  total: number;
}
