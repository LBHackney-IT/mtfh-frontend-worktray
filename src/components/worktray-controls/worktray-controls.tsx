import React, { useCallback, useContext } from "react";

import { FormGroup, Select } from "@mtfh/common/lib/components";

import { WorktrayContext } from "../../context/worktray-context";
import { locale } from "../../services";
import { LimitOptions, TimePeriodOptions } from "../../types";

import "./styles.scss";

const { timePeriodOptions, limitOption } = locale.components.controls;

export const WorktrayControls = (): JSX.Element | null => {
  const {
    state: { pageSize, timePeriod },
    dispatch,
  } = useContext(WorktrayContext);

  const handleTimePeriod = useCallback(
    (option: TimePeriodOptions) => {
      dispatch({ type: "TIME_PERIOD", payload: option });
    },
    [dispatch],
  );

  const handleLimit = useCallback(
    (limit: LimitOptions) => {
      dispatch({ type: "LIMIT", payload: limit });
    },
    [dispatch],
  );

  return (
    <div className="mtfh-worktray-controls">
      <div className="mtfh-worktray-controls__controls">
        <FormGroup id="time-period" label="Show:">
          <Select
            data-testid="mtfh-worktray-time-period"
            onChange={(e) => handleTimePeriod(e.currentTarget.value as TimePeriodOptions)}
            value={timePeriod}
          >
            {Object.values(TimePeriodOptions)
              .filter((value) => typeof value === "number" || value === "")
              .map((value) => (
                <option key={value} value={value}>
                  {timePeriodOptions[value] || `All of ${new Date().getFullYear()}`}
                </option>
              ))}
          </Select>
        </FormGroup>

        <FormGroup id="limit" label="">
          <Select
            data-testid="mtfh-worktray-limit"
            onChange={(e) => handleLimit(Number(e.currentTarget.value))}
            value={pageSize}
          >
            {(Object.values(LimitOptions) as number[])
              .filter((value) => typeof value === "number")
              .map((value) => (
                <option key={value} value={value}>
                  {limitOption(value)}
                </option>
              ))}
          </Select>
        </FormGroup>
      </div>
    </div>
  );
};
