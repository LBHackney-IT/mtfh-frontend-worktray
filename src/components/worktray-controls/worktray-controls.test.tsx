import React from "react";

import { render } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { WorktrayProvider } from "../../context/worktray-context";
import { LimitOptions, TimePeriodOptions } from "../../types";
import { WorktrayControls } from "./worktray-controls";

describe("worktray-controls-component", () => {
  test("WorktrayControls renders correctly", () => {
    const { container } = render(
      <WorktrayProvider initial={{}}>
        <WorktrayControls />
      </WorktrayProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test("WorktrayControls changes time period", async () => {
    render(
      <WorktrayProvider initial={{}}>
        <WorktrayControls />
      </WorktrayProvider>,
    );

    const timePeriod = screen.getByTestId(
      "mtfh-worktray-time-period",
    ) as HTMLSelectElement;

    await userEvent.selectOptions(timePeriod, String(TimePeriodOptions.DAYS_60));
    expect(timePeriod.value).toBe(String(TimePeriodOptions.DAYS_60));
  });

  test("WorktrayControls changes limit", async () => {
    render(
      <WorktrayProvider initial={{}}>
        <WorktrayControls />
      </WorktrayProvider>,
    );

    const limit = screen.getByTestId("mtfh-worktray-limit") as HTMLSelectElement;

    await userEvent.selectOptions(limit, String(LimitOptions.MEDIUM));
    expect(limit.value).toBe(String(LimitOptions.MEDIUM));
  });
});
