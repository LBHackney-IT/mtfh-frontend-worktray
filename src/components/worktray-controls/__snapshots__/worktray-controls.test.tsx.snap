// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worktray-controls-component WorktrayControls renders correctly 1`] = `
<div>
  <div
    class="mtfh-worktray-controls"
  >
    <div
      class="mtfh-worktray-controls__controls"
    >
      <div
        class="govuk-form-group lbh-form-group"
        id="time-period"
      >
        <label
          class="govuk-label lbh-label"
          for="time-period-field"
        >
          Show:
          
        </label>
        <select
          class="govuk-select lbh-select"
          data-testid="mtfh-worktray-time-period"
          id="time-period-field"
        >
          <option
            value="30"
          >
            Last 30 days
          </option>
          <option
            value="60"
          >
            Last 60 days
          </option>
          <option
            value="3"
          >
            Last 3 months
          </option>
          <option
            value="6"
          >
            Last 6 months
          </option>
          <option
            value="12"
          >
            Last 12 months
          </option>
          <option
            value=""
          >
            All of 2022
          </option>
        </select>
      </div>
      <div
        class="govuk-form-group lbh-form-group"
        id="limit"
      >
        <label
          class="govuk-label lbh-label"
          for="limit-field"
        >
          
          
        </label>
        <select
          class="govuk-select lbh-select"
          data-testid="mtfh-worktray-limit"
          id="limit-field"
        >
          <option
            value="10"
          >
            10 items
          </option>
          <option
            value="25"
          >
            25 items
          </option>
          <option
            value="50"
          >
            50 items
          </option>
          <option
            value="100"
          >
            100 items
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
